keps/sig-node/3008-qos-class-resources/README.md

```diff
@@ -0,0 +1,2248 @@
+    [PodSandboxConfig](#podsandboxconfig) messages
+  - updates Pod status to reflect the assignment of QoS-class resources, based
+    on pod/container status reveived from the runtime
+  - implements admission handler that validates the availability of QoS-class resources
+  - pod eviction as a possible [future improvement](#kubelet-initiated-pod-eviction)
+- scheduler
+  - simple node filtering based on the availability of QoS-class resources in
+    the first implementation phase (with possible
+    [future improvements](#scheduler-improvements))
+  - pod priority based eviction is in effect
+- quota, limits and defaults are [future work](#future-work) items
+
+
+### CRI API
+
+The following additions to the CRI protocol are suggested.
+
+#### Implicit defaults
+
+Any defaults imposed by the runtime, i.e. for resources that the user didn't
+request anything, must be reflected in the PodSandboxStatus and ContainerStatus
+messages. This information (about implicit defaults) can change during the
+lifetime of the Pod e.g. because of re-configuration of the runtime:
+
+- new QoS resource types might be added in which case a Pod/Container is
+  assigned to the default class of this new resource and this information must
+  be reflected in PodSandboxStatus/ContainerStatus
+- the default class of an existing QoS resource type might change and the
+  Pod/Container migrated to this class and this information must be reflected
+  in PodSandboxStatus/ContainerStatus
+
+An empty value (empty string) denotes "system default" i.e. the runtime did not
+enforce any QoS for this specific type of QoS-class resource. An example can be
+for example Linux cgroup controls where "system default" would mean that the
+runtime did not enforce any changes so they will be set to the default values
+determined by the system configuration.
+
+#### ContainerConfig
+
+The `ContainerConfig` message will be supplemented with new `qos_resources`
+field, providing per-container setting for QoS-class resources. This will be
+used in `CreateContainerRequest` to communicate the container-level QoS-class
+resource assignments to the runtime.
+
+```diff
+ message ContainerConfig {
+
+ ...
+     // Configuration specific to Linux containers.
+     LinuxContainerConfig linux = 15;
+     // Configuration specific to Windows containers.
+     WindowsContainerConfig windows = 16;
++
++    // Configuration of QoS resources.
++    repeated ContainerQOSResource qos_resources = 17;
+ }
+
++// ContainerQOSResource specifies the assignment of one QoS resource for a
++// container.
++message ContainerQOSResource {
++    // Name of the QoS resource.
++    string name = 1;
++    // Name of the class.
++    string class = 2;
++}
+```
+
+`<<[UNRESOLVED @haircommander @mrunalp]>>`
+
+Instead of separate types (ContainerQOSResource ans PodQOSResource) use shared
+type (say QOSResource) for assignment/status of both container and pod level
+resources.
```

Comment:
I'm of the opinion we should compress them if we don't expect any additional metadata to be needed for either.

nit: and*

Reply:
ah I see https://github.com/kubernetes/enhancements/pull/3004/files#r1191541425 now. sounds like @thockin and I disagree on reusing

Reply:
@mrunalp ^